### ISSUE TITLE
"Invalid Input!" occurs at fully connected layer during training

### ISSUE THREAD

**Issue by gschmidt958** [cbovar/ConvNetSharp#153]
This is my first network with ConvNetSharp so it's likely I'm doing something fundamentally wrong.  After a few iterations during training, training crashes with "Invalid Input!" when the first fully connected layer 1x1x1000 is reached.  Looking at the storage array, I notice that the array is filled with "Infinity" and "-Infinity".  The input to the net is just some some contrived test data which alternates between 0's and 1's.  Below is the crash and my network topology.  Thanks for any insights on this!

![image](https://user-images.githubusercontent.com/5486583/80040059-3e7d9f80-84c7-11ea-92d3-6e22d4243921.png)

```csharp
            net.AddLayer(new InputLayer(7, 9, 6));

            // Convolution layer #1 2x2
            net.AddLayer(new ConvLayer(2, 2, 6) { Stride = 1 });
            // Result after applying filter is (6 x 8) x 6_filters = 288 elements

            // declare a ReLU (rectified linear unit non-linearity)
            net.AddLayer(new ReluLayer());

            // Convolution layer #2 2x2
            net.AddLayer(new ConvLayer(2, 2, 6) { Stride = 1 });
            // Result after applying filter is (5 x 7) x 6_filters = 210 elements

            // declare a ReLU (rectified linear unit non-linearity)
            net.AddLayer(new ReluLayer());

            // Fully connected layer #1.
            net.AddLayer(new FullyConnLayer(1000));

            // Fully connected layer #2
            net.AddLayer(new FullyConnLayer(500));

            // Fully connected layer #3 - input to softmax.
            net.AddLayer(new FullyConnLayer(1));

            // declare the linear classifier on top of the previous hidden layer
            net.AddLayer(new SoftmaxLayer(1));

            // Construct a trainer for our network
            var trainer = new SgdTrainer(net) { LearningRate = 0.01, Momentum = 0.0, BatchSize = 1 };

            var input = BuilderInstance.Volume.SameAs(new Shape(7, 9, 6));
            var label = BuilderInstance.Volume.SameAs(new Shape(1, 1, 1));

            var evenInput = BuilderInstance.Volume.SameAs(new Shape(7, 9, 6));
            var oddInput = BuilderInstance.Volume.SameAs(new Shape(7, 9, 6));
            MakeEven(evenInput);
            MakeOdd(oddInput);

            for (int i = 0; i < 100; i++)
            {
                Console.WriteLine("iteration: " + i);

                label.Set(0, 1.0);
                trainer.Train(oddInput, label);

                label.Set(0, 0.0);
                trainer.Train(evenInput, label);
            }
```

**Comment by cbovar** [comment]
Hello,
It seems your gradients are exploding. Try to reduce the learning rate ? (e.g. 0.001 or 0.0001)

**Comment by gschmidt958** [comment]
Hello,
Thanks, but reducing the learning rate just pushed the problem down to a later iteration.  Right now, I'd just like to try to get to a working starting point and work up from there, so I ended up simplifying the network by going from 3 to 2 fully connected layers.  That seems to eliminate the gradient explosion problem.  However, after creating two test cases initialized with two different constant values, and running through 1000 iterations, I get exactly the same classification probabilities for both patterns.  Initially, I had a single class, but then changed it to two (thinking that might somehow help, but of course it didn't).  I'm sure I must be doing something very fundamentally wrong (most likely a head slap looming).  If you would be willing to take a look at my code (below), that would be much appreciated.  Thanks!

Output
![image](https://user-images.githubusercontent.com/5486583/80270080-22bef880-8683-11ea-9118-6a05d025f2d7.png)

```csharp
using System;
using ConvNetSharp.Core;
using ConvNetSharp.Core.Layers.Double;
using ConvNetSharp.Core.Training.Double;
using ConvNetSharp.Volume;
using ConvNetSharp.Volume.Double;

namespace ClusterNN
{
    internal class Program
    {
        private static void Main()
        {
            var net = new Net<double>();

            net.AddLayer(new InputLayer(7, 9, 6));

            // Convolution layer #1 2x2
            net.AddLayer(new ConvLayer(2, 2, 6) { Stride = 1 });
            // Result after applying filter is (6 x 8) x 6_filters = 288 elements

            // declare a ReLU (rectified linear unit non-linearity)
            net.AddLayer(new ReluLayer());

            // Convolution layer #2 2x2
            net.AddLayer(new ConvLayer(2, 2, 6) { Stride = 1 });
            // Result after applying filter is (5 x 7) x 6_filters = 210 elements

            // declare a ReLU (rectified linear unit non-linearity)
            net.AddLayer(new ReluLayer());

            // Fully connected layer #1.
            net.AddLayer(new FullyConnLayer(100));
            
            // Fully connected layer #2
            net.AddLayer(new FullyConnLayer(2));

            // declare the linear classifier on top of the previous hidden layer
            net.AddLayer(new SoftmaxLayer(2));

            // Construct a trainer for our network
            var trainer = new SgdTrainer(net) { LearningRate = 0.001, Momentum = 0.0, BatchSize = 1 };

            var evenInput = BuilderInstance.Volume.SameAs(new Shape(7, 9, 6));
            var oddInput = BuilderInstance.Volume.SameAs(new Shape(7, 9, 6));

            MakeEvenOddVolumes(evenInput, oddInput);

            var label0_2 = BuilderInstance.Volume.SameAs(new Shape(1, 1, 2));
            label0_2.Set(0, 0, 0, 0.0);
            label0_2.Set(0, 0, 1, 1.0);

            var label1_2 = BuilderInstance.Volume.SameAs(new Shape(1, 1, 2));
            label1_2.Set(0, 0, 0, 1.0);
            label1_2.Set(0, 0, 1, 0.0);

            for (int i = 0; i < 1000; i++)
            {
                Console.WriteLine("iteration: " + i);

                trainer.Train(evenInput, label0_2);
                trainer.Train(oddInput, label1_2);
            }

            // forward a data point through the network
            var vscore = net.Forward(evenInput);
            Console.WriteLine("probability that x is even: " + vscore.Get(0, 0, 0) + " " + vscore.Get(0, 0, 1));

            var vscore2 = net.Forward(oddInput);
            Console.WriteLine("probability that x is odd: " + vscore2.Get(0, 0, 0) + " " +  vscore2.Get(0, 0, 1));

        }

        private static void MakeEvenOddVolumes(Volume<double> evenVolume, Volume<double> oddVolume)
        {
            int counter = 0;
            for (int i = 0; i < 7; i++)
            {
                for (int j = 0; j < 9; j++)
                {
                    for (int k = 0; k < 6; k++)
                    {
                        evenVolume.Set(i, j, k, 100.0);
                        oddVolume.Set(i, j, k, 11);
                        counter++;
                    }
                }
            }
        }
    }
}
```

**Comment by cbovar** [comment]
I noticed the loss is not changing much from one iteration to the other so I played with the learning rate:
Increasing learning rate to `1.0` seems to solve this issue.
Another way would be changing `SgdTrainer `by `AdamTrainer`.

![image](https://user-images.githubusercontent.com/5098385/80557737-81310180-8a0a-11ea-8e00-68cc0dd6a0ac.png)

I'm not sure why such a high learning rate is needed. It means that the computed gradients are very small.
